### PR TITLE
Raven db 15994 spatial shapes

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs [hunk 1]
```diff
@@ -230,6 +231,72 @@ public static void WriteFacetedQueryResult(this BlittableJsonTextWriter writer,
...
+                writer.WritePropertyName(nameof(circle.Units));
+                writer.WriteString(circle.Units.ToString());
+            }
+            
+            else if (result.ShapeType == SpatialShape.Polygon)
+            {
+                var polygon = result as Polygon;
+                writer.WriteArray(context, nameof(polygon.Vertices), polygon.Vertices, 
+                    (w, c, point) => w.WriteSpatialPointResult(c, point));
+            }
```
Comment (ml054) [inline src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:278]: missing else and throw here?

File: src/Raven.Studio/bower.json [hunk 2]
```diff
@@ -39,7 +39,9 @@
     "forge": "^0.6.49",
     "jszip": "^3.1.5",
     "jszip-utils": "^0.0.2",
-    "diff": "https://cdnjs.cloudflare.com/ajax/libs/jsdiff/4.0.1/diff.min.js"
+    "diff": "https://cdnjs.cloudflare.com/ajax/libs/jsdiff/4.0.1/diff.min.js",
+    "leaflet": "^1.7.1",
```
Comment (ml054) [inline src/Raven.Studio/bower.json:43]: didn't we already add that?

File: src/Raven.Studio/typescript/common/generalUtils.ts [hunk 3]
```diff
@@ -642,6 +649,20 @@ class genUtils {
         div.remove();
         return html;
     }
+    
+    static flattenObj(obj: any, parentKey: string, res = {}) {
+        for (let key in obj){
```
Comment (ml054) [inline src/Raven.Studio/typescript/common/generalUtils.ts:654]: missing space

File: src/Raven.Studio/typescript/common/generalUtils.ts [hunk 4]
```diff
@@ -642,6 +649,20 @@ class genUtils {
...
+            const propName = parentKey ? parentKey + "." + key : key;
+            const value = obj[key];
+
+            if (typeof value === "object"){
```
Comment (ml054) [inline src/Raven.Studio/typescript/common/generalUtils.ts:658]: missing space

File: src/Raven.Studio/typescript/models/database/query/spatialPolygonModel.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+class spatialPolygonModel {    
+    
+    static readonly colors = ["#38D6CC", "#12D366", "#95E716"];
+    
+    vertices: Array<[number, number]>;
+    
+    constructor(polygon: Raven.Client.Documents.Indexes.Spatial.Polygon) {
+       
+       this.vertices = polygon.Vertices.map(v => {
+           return [v.Latitude, v.Longitude];
```
Comment (ml054) [inline src/Raven.Studio/typescript/models/database/query/spatialPolygonModel.ts:10]: no need to return v => [v.Latitude, v.Longitude]

File: src/Raven.Studio/typescript/viewmodels/database/query/query.ts [hunk 6]
```diff
@@ -1185,6 +1197,64 @@ class query extends viewModelBase {
...
+                    const item = queryResults.items[i];
+                    const flatItem = generalUtils.flattenObj(item, "");
+                    
+                    const latitudeValue = _.get(flatItem, latitudeProperty) as number;
+                    const longitudeValue = _.get(flatItem, longitudeProperty) as number;
+                    
+                    if (latitudeValue != null && longitudeValue != null) {
+                        const point: geoPointInfo = { Latitude: latitudeValue, Longitude: longitudeValue, PopupContent: item };
+                        pointsArray.push(point);
+                        this.totalNumberOfMarkers(this.totalNumberOfMarkers() + 1);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/query/query.ts:1228]: please don't update observable in for loop - just count this using variable and update one time.

File: src/Raven.Studio/typescript/viewmodels/database/query/spatialQueryMap.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            const textHtml = `<div>
+                                    <h4>Document: ${genUtils.escapeHtml(doc.getId())}</h4>
+                                    <hr>
+                                    <pre>${text}</pre>
+                              </div>`;
+            return textHtml;
+        }
+        
+        const dataLayers: Control.LayersObject = {};
+        let markersGroups: MarkerClusterGroup[] = [];
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/query/spatialQueryMap.ts:54]: const?

**Comment by Danielle9897** [comment]
will use previous pr https://github.com/ravendb/ravendb/pull/11534/commits

File: src/Raven.Studio/typescript/models/database/query/spatialPolygonModel.ts (hunk 5 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/query/spatialPolygonModel.ts:10]: done

File: src/Raven.Studio/typescript/viewmodels/database/query/spatialQueryMap.ts (hunk 7 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/query/spatialQueryMap.ts:54]: done

File: src/Raven.Studio/typescript/common/generalUtils.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/common/generalUtils.ts:658]: done

File: src/Raven.Studio/typescript/common/generalUtils.ts (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/common/generalUtils.ts:654]: done

File: src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:278]: done

File: src/Raven.Studio/typescript/viewmodels/database/query/query.ts (hunk 6 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/query/query.ts:1228]: done

**Pull request closed without merging**